File: backend/python/app/telemetry/event_buffer.py
"""Thread-safe, bounded buffer for product-behaviour events awaiting shipment.

Unlike logs, events are individually meaningful (not aggregated). Drained by the
MetricsPusher tick. Bounded so a burst can't exhaust memory.
"""

import threading
from datetime import datetime, timezone
from typing import List, Optional

MAX_EVENTS = 5000


class EventBuffer:
    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._events: List[dict] = []

    def enqueue(self, event: dict) -> None:
        with self._lock:
            if len(self._events) >= MAX_EVENTS:
                return  # drop on overflow
            self._events.append(event)

    def drain(self) -> List[dict]:
        with self._lock:
            events = self._events
            self._events = []
            return events

    def size(self) -> int:
        with self._lock:
            return len(self._events)


event_buffer = EventBuffer()


def record_event(event: str, props: Optional[dict] = None) -> None:
    """Record a product-behaviour event (search_performed, agent_run, ...).

    Props may carry identity (email/org/user) and dimensions (connector/feature)
    but must NOT carry document/query content — the collector drops
    content-smelling keys as a backstop.
    """
    event_buffer.enqueue(
        {
            "event": event,
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "props": props or {},
        }
    )

File: backend/nodejs/apps/src/libs/services/telemetry/event-buffer.ts
export interface BufferedEvent {
  event: string;
  timestamp: string;
  props?: Record<string, unknown>;
}

const MAX_EVENTS = 5000;

class EventBuffer {
  private events: BufferedEvent[] = [];

  enqueue(event: BufferedEvent): void {
    if (this.events.length >= MAX_EVENTS) return;
    this.events.push(event);
  }

  drain(): BufferedEvent[] {
    const drained = this.events;
    this.events = [];
    return drained;
  }

  size(): number {
    return this.events.length;
  }
}

export const eventBuffer = new EventBuffer();

export function recordEvent(
  event: string,
  props?: Record<string, unknown>,
): void {
  eventBuffer.enqueue({ event, timestamp: new Date().toISOString(), props });
}
